store/store_download.go

```diff
@@ -80,14 +82,79 @@ func init() {
 }
 
 // Deltas enabled by default on classic, but allow opting in or out on both classic and core.
-func useDeltas() bool {
-	// only xdelta3 is supported for now, so check the binary exists here
-	// TODO: have a per-format checker instead
-	if _, err := getXdelta3Cmd(); err != nil {
+func (s *Store) useDeltas() (use bool) {
+	s.xdeltaCheckLock.Lock()
+	defer s.xdeltaCheckLock.Unlock()
+
+	// check the cached value if available
+	if s.shouldUseDeltas != nil {
+		return *s.shouldUseDeltas
+	}
+
+	defer func() {
+		// cache whatever value we return for next time
+		s.shouldUseDeltas = &use
+	}()
+
+	// check if deltas were disabled by the environment
+	if !osutil.GetenvBool("SNAPD_USE_DELTAS_EXPERIMENTAL", true) {
+		// then the env var is explicitly false, we can't use deltas
+		logger.Debugf("delta usage disabled by environment variable")
 		return false
 	}
 
-	return osutil.GetenvBool("SNAPD_USE_DELTAS_EXPERIMENTAL", true)
+	// TODO: have a per-format checker instead, we currently only support
+	// xdelta3 as a format for deltas
+
+	// check if the xdelta3 config command works from the system snap
+	cmd, err := commandFromSystemSnap("/usr/bin/xdelta3", "config")
+	if err == nil {
+		// we have a xdelta3 from the system snap, make sure it works
+		if runErr := cmd.Run(); runErr == nil {
+			// success using the system snap provided one, setup the callback to
+			// use the cmd we got from CommandFromSystemSnap, but with a small
+			// tweak - this cmd to run xdelta3 from the system snap will likely
+			// have other arguments and a different main exe usually, so
+			// use it exactly as we got it from CommandFromSystemSnap,
+			// but drop the last arg which we know is "config"
+			exe := cmd.Path
+			args := cmd.Args[:len(cmd.Args)-1]
+			env := cmd.Env
+			dir := cmd.Dir
+			s.xdelta3CmdFunc = func(xDelta3args ...string) *exec.Cmd {
+				return &exec.Cmd{
+					Path: exe,
+					Args: append(args, xDelta3args...),
+					Env:  env,
```

Comment:
do we need to copy Env and Dir? afaiu the snapdtool helper itself uses exec.Command which doesn't set those

Reply:
No, but I thought it good practice since snapdtool.CommandFromSystemSnap could some day return more things in the Cmd and we may not catch that here when we make that change. I can remove this if you want since snapdtool does not currently set those as you pointed out